### ISSUE TITLE
bug: ContentLength not set on Java SDK 2.0 when performing S3 get-object

### ISSUE THREAD

**Issue by dominic-jones** [localstack/localstack#9294]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

This sounds like a regression of this issue: https://github.com/localstack/localstack/issues/631 . The description there sounds the same as what we're experiencing.

After upgrading to localstack 2.2 (and persists to 2.3), our AWS SDK now receives a null ContentLength returned on getting an S3 object.

When performing the same check with the AWS CLI, the ContentLength is returned correctly.

I compared some exact headers using WireShark.

This compares WireShark outputs: left AWS SDK on localstack 2.1, right is AWS SDK on localstack 2.2:
![aws cli and aws sdk localstack 2-2](https://github.com/localstack/localstack/assets/5888934/01e18a9d-6b02-4c86-8028-4e36aecfe63c)

This is both localstack 2.2, left is the AWS CLI, right is the AWS SDK:
![aws sdk localstack 2-1 and 2-2](https://github.com/localstack/localstack/assets/5888934/dcd685e8-dfd3-419e-8040-43b56fa130c1)

### Expected Behavior

When getting an object from S3, the ContentLength should be returned to the AWS SDK.

### How are you starting LocalStack?

With a docker-compose file

#### Docker commands

I don't have access to these, sorry. There's a layer of company scripting in front that obscures the exact command.

#### AWS SDK Execution 

```kotlin
        val request = GetObjectRequest.builder()
            .bucket(actualS3BucketName)
            .key(s3Key)
            .apply { range?.let { range(HttpRange.toString(listOf(it))) } }
            .build()
        val respTransformer = SyncGetObjectResponseTransformer()
        return this.objectSyncClient
            .getObject(request, respTransformer)
```

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

```sh
aws s3api get-object --bucket s3-objects \
  --key "$key" \
  output-file-name \
  --endpoint-url http://localhost:4566 --profile localstack
```

### Environment

```markdown
OS: Ubuntu 22.04
AWS SDK: 2.20.159
localstack: 2.2.0 (also 2.3.0 using v2 provider)
```

**Comment by bentsku** [comment]
Hello @dominic-jones and thanks for your report!

I think this is a continuation of #8845, this might not have fixed the issue entirely.

I can see in the diff between 2.1 and 2.2 that the URL to access LocalStack has changed too, and I think with 2.2 and 2.3 if you access LocalStack S3 with `http://localhost:4566` and accessing the bucket in "path style" (bucket part as the path and not the host), you shouldn't encounter the issue with `ContentLength` missing. This is indeed only a workaround, and we will work to fix this issue. Thanks!

**Comment by dominic-jones** [comment]
Interesting! Good catch. The WireShark report shows it's calling on the path-style syntax, even though my SDK logs show a different host: ```
2023-10-05 16:49:20.099 DEBUG 497401 --- [s.test runner#6] software.amazon.awssdk.request           : Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=http, host=s3-objects.s3.localhost.localstack.cloud, port=4566, encodedPath=/9245759a-9281-4402-82a6-6470d85e342e/97c7eae4-5441-4482-a758-ef72c42564bc/05351694-4b7a-4ec6-a3ed-3916a1b7a14a/PvnCE8u01Z/42, headers=[amz-sdk-invocation-id, Content-Length, Content-Type, Expect, User-Agent, x-amz-checksum-sha1, x-amz-tagging], queryParameters=[]) contextMap={} 
```
I'll investigate if there's any translation occurring in our system. Thanks for the direction. Good luck with the fix!

**Comment by bentsku** [comment]
Ah, maybe you intercepted our own call, for now, to support virtual host requests, we forward the rewritten request to ourselves. Maybe that’s what has been picked up?

**Comment by bentsku** [comment]
I'm currently unable to reproduce the issue and I might be missing something. Would there be any way to share a small reproducible sample so that I would be able to reproduce and try to fix the issue? Thanks!

**Comment by bentsku** [comment]
I can share how I've tried to reproduce, and the LocalStack logs, so maybe you can help me reproduce:

the Java Snippet with AWS SDK v2 (v2.20.159)
```java
GetObjectRequest get_request = GetObjectRequest.builder()
        .bucket(bucketName)
        .key(key)
        .build();

ResponseInputStream<GetObjectResponse> responseInputStream = client.getObject(get_request);
GetObjectResponse response = responseInputStream.response();
Long contentLength = response.contentLength();
assert contentLength == 100;
```

LocalStack logs for the request:
```bash
2023-10-05T20:53:09.548 DEBUG --- [   asgi_gw_0] l.aws.serving.wsgi         : GET integration-bucket.s3.localhost.localstack.cloud:4566/testObjectKey_5f1gUa
2023-10-05T20:53:09.548 DEBUG --- [   asgi_gw_0] l.services.s3.virtual_host : Rewritten original host url: http://integration-bucket.s3.localhost.localstack.cloud:4566/testObjectKey_5f1gUa to path-style url: http://s3.localhost.localstack.cloud:4566/integration-bucket/testObjectKey_5f1gUa
2023-10-05T20:53:09.552 DEBUG --- [   asgi_gw_1] l.aws.serving.wsgi         : GET s3.localhost.localstack.cloud:4566/integration-bucket/testObjectKey_5f1gUa
2023-10-05T20:53:09.555 DEBUG --- [   asgi_gw_1] l.aws.protocol.serializer  : Determined accept type (None) is not supported by this serializer. Using default of this serializer: application/xml
2023-10-05T20:53:09.556  INFO --- [   asgi_gw_1] localstack.request.aws     : AWS s3.GetObject => 200; 000000000000/us-east-1; GetObjectRequest({'Bucket': 'integration-bucket', 'IfMatch': None, 'IfModifiedSince': None, 'IfNoneMatch': None, 'IfUnmodifiedSince': None, 'Key': 'testObjectKey_5f1gUa', 'Range': None, 'ResponseCacheControl': None, 'ResponseContentDisposition': None, 'ResponseContentEncoding': None, 'ResponseContentLanguage': None, 'ResponseContentType': None, 'ResponseExpires': None, 'VersionId': None, 'SSECustomerAlgorithm': None, 'SSECustomerKey': None, 'SSECustomerKeyMD5': None, 'RequestPayer': None, 'PartNumber': None, 'ExpectedBucketOwner': None, 'ChecksumMode': None}, headers={'User-Agent': 'aws-sdk-java/2.20.159 Mac_OS_X/10.16 OpenJDK_64-Bit_Server_VM/13.0.2+8 Java/13.0.2 scala/2.12.10 kotlin/1.7.10-release-333(1.7.10) vendor/Oracle_Corporation io/sync http/Apache cfg/retry-mode/standard ft/s3-transfer', 'accept-encoding': 'identity', 'Accept': '*/*', 'Connection': 'Keep-Alive', 'Host': 's3.localhost.localstack.cloud:4566', 'Content-Length': '0', 'Amz-Sdk-Invocation-Id': '6bffb97a-61cf-4dac-e75c-43855675a228', 'Amz-Sdk-Request': 'attempt=1; max=3', 'Authorization': 'AWS4-HMAC-SHA256 Credential=access/20231005/us-east-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;host;x-amz-content-sha256;x-amz-date;x-amz-te, Signature=7b85e2f9fea5facb87f38c47354cfb16bfa708187f169da98e1ad94a6b49b898', 'X-Amz-Content-Sha256': 'UNSIGNED-PAYLOAD', 'X-Amz-Date': '20231005T175309Z', 'X-Amz-Te': 'append-md5', 'x-localstack-tgt-api': 's3', 'X-S3-Vhost-Forwarded-For': 'integration-bucket.s3.localhost.localstack.cloud:4566', 'X-Forwarded-For': '127.0.0.1', 'x-moto-account-id': '000000000000'}); GetObjectOutput({'LastModified': datetime.datetime(2023, 10, 5, 17, 53, 9, tzinfo=tzutc()), 'ContentLength': 100, 'ETag': '"e5272a0c8f8392c3c75c0449f06ced7a"', 'ContentType': 'application/octet-stream', 'Metadata': {}, 'StatusCode': 200, 'Body': StreamedBytes(100), 'AcceptRanges': 'bytes'}, headers={'Content-Type': 'application/octet-stream', 'accept-ranges': 'bytes', 'Last-Modified': 'Thu, 05 Oct 2023 17:53:09 GMT', 'Content-Length': '100', 'ETag': '"e5272a0c8f8392c3c75c0449f06ced7a"', 'x-amz-request-id': '33e47560-edb0-43ba-9ffd-88068ebcc4c2', 'x-amz-id-2': 's9lzHYrFp76ZVxRcpX9+5cjAnEH2ROuNkd2BHfIa6UkFVdtjf5mKR3/eTPFvsiP/XV/VLi31234='})
2023-10-05T20:53:09.557  INFO --- [   asgi_gw_0] localstack.request.http    : GET /testObjectKey_5f1gUa => 200; Request(b'', headers={'Host': 'integration-bucket.s3.localhost.localstack.cloud:4566', 'amz-sdk-invocation-id': '6bffb97a-61cf-4dac-e75c-43855675a228', 'amz-sdk-request': 'attempt=1; max=3', 'Authorization': 'AWS4-HMAC-SHA256 Credential=access/20231005/us-east-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;host;x-amz-content-sha256;x-amz-date;x-amz-te, Signature=7b85e2f9fea5facb87f38c47354cfb16bfa708187f169da98e1ad94a6b49b898', 'User-Agent': 'aws-sdk-java/2.20.159 Mac_OS_X/10.16 OpenJDK_64-Bit_Server_VM/13.0.2+8 Java/13.0.2 scala/2.12.10 kotlin/1.7.10-release-333(1.7.10) vendor/Oracle_Corporation io/sync http/Apache cfg/retry-mode/standard ft/s3-transfer', 'x-amz-content-sha256': 'UNSIGNED-PAYLOAD', 'X-Amz-Date': '20231005T175309Z', 'x-amz-te': 'append-md5', 'Connection': 'Keep-Alive', 'x-localstack-tgt-api': 's3'}); Response(StreamingBody(unknown), headers={'Content-Type': 'application/octet-stream', 'accept-ranges': 'bytes', 'Last-Modified': 'Thu, 05 Oct 2023 17:53:09 GMT', 'Content-Length': '100', 'ETag': '"e5272a0c8f8392c3c75c0449f06ced7a"', 'x-amz-request-id': '33e47560-edb0-43ba-9ffd-88068ebcc4c2', 'x-amz-id-2': 's9lzHYrFp76ZVxRcpX9+5cjAnEH2ROuNkd2BHfIa6UkFVdtjf5mKR3/eTPFvsiP/XV/VLi31234='})
```

And some screenshots from the debugger:
<img width="847" alt="Screenshot 2023-10-05 at 16 39 59" src="https://github.com/localstack/localstack/assets/42031100/bfb4a893-88fb-4c2f-bac1-772ca579190c">
<img width="510" alt="Screenshot 2023-10-05 at 16 40 16" src="https://github.com/localstack/localstack/assets/42031100/f82a026e-69a0-44f4-a304-321f8c9ac167">

Is there an additional step I need to take to reproduce the issue?

**Comment by dominic-jones** [comment]
From your own logs there, it looks like you're accessing the bucket through the domain-style access, which is promising that a content-length is returned! Could well be, as you said, that the path-style access is the part that's broken.

Then again, my JVM/slf4j logging also shows domain-style, but intercepting with WireShark showed it was going through path-style. I'm investigating to see why this change occurs.

**Comment by dominic-jones** [comment]
I believe this is the relevant part of the logs:

```
2023-10-05T22:18:50.828  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS s3.GetObject => 200; GetObjectRequest({'Bucket': 's3-objects', 'IfMatch': None, 'IfModifiedSince': None, 'IfNoneMatch': None, 'IfUnmodifiedSince': None, 'Key': 'db889fb4-ec6d-4ba6-9b32-7d8658c96e49/dceb11d3-246b-4522-80cd-eab859078a87/76a2e57e-360a-47b6-8cb2-c4d4fc9a661e/M1tIcruD2Z/98', 'Range': None, 'ResponseCacheControl': None, 'ResponseContentDisposition': None, 'ResponseContentEncoding': None, 'ResponseContentLanguage': None, 'ResponseContentType': None, 'ResponseExpires': None, 'VersionId': None, 'SSECustomerAlgorithm': None, 'SSECustomerKey': None, 'SSECustomerKeyMD5': None, 'RequestPayer': None, 'PartNumber': None, 'ExpectedBucketOwner': None, 'ChecksumMode': None}, headers={'User-Agent': 'aws-sdk-java/2.20.159 Linux/6.4.6-76060406-generic OpenJDK_64-Bit_Server_VM/17.0.7+7-LTS Java/17.0.7 scala/2.12.10 kotlin/1.9.10-release-459 vendor/Amazon.com_Inc. io/async http/NettyNio cfg/retry-mode/legacy', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Host': 's3.localhost.localstack.cloud:4566', 'Content-Length': '0', 'Amz-Sdk-Invocation-Id': 'ebe65d10-9ef6-1702-7568-d964262f3905', 'Amz-Sdk-Request': 'attempt=1; max=4', 'Authorization': 'AWS4-HMAC-SHA256 Credential=arbitrary-value-is-required-for-localstack/20231005/ap-south-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;host;x-amz-content-sha256;x-amz-date;x-amz-te, Signature=bbb772c4b64742d6a3c7f2129128cf56191507fecfb15da8ea0f9aac27125dd5', 'X-Amz-Content-Sha256': 'UNSIGNED-PAYLOAD', 'X-Amz-Date': '20231005T221850Z', 'X-Amz-Te': 'append-md5', 'x-localstack-tgt-api': 's3', 'X-S3-Vhost-Forwarded-For': 's3-objects.s3.localhost.localstack.cloud:4566', 'X-Forwarded-For': '192.168.80.1', 'x-moto-account-id': '000000000000'}); GetObjectOutput({'LastModified': datetime.datetime(2023, 10, 5, 22, 18, 50, tzinfo=tzutc()), 'ContentLength': 100, 'ETag': '"fbaff7b2f613227d604f3e89f517d379"', 'VersionId': '55460f66-2508-4ad5-9404-08a6b97b5e34', 'ContentType': 'application/octet-stream', 'Metadata': {}, 'TagCount': 3, 'StatusCode': 200, 'Body': StreamedBytes(100), 'AcceptRanges': 'bytes'}, headers={'Content-Type': 'application/octet-stream', 'accept-ranges': 'bytes', 'Last-Modified': 'Thu, 05 Oct 2023 22:18:50 GMT', 'Content-Length': '100', 'ETag': '"fbaff7b2f613227d604f3e89f517d379"', 'x-amz-version-id': '55460f66-2508-4ad5-9404-08a6b97b5e34', 'x-amz-tagging-count': '3', 'x-amz-request-id': '0a2a9574-467d-40f9-9cd4-d203c3eceb2c', 'x-amz-id-2': 's9lzHYrFp76ZVxRcpX9+5cjAnEH2ROuNkd2BHfIa6UkFVdtjf5mKR3/eTPFvsiP/XV/VLi31234='})
2023-10-05T22:18:50.830  INFO --- [   asgi_gw_1] localstack.request.http    : GET /db889fb4-ec6d-4ba6-9b32-7d8658c96e49/dceb11d3-246b-4522-80cd-eab859078a87/76a2e57e-360a-47b6-8cb2-c4d4fc9a661e/M1tIcruD2Z/98 => 200; Request(b'', headers={'Host': 's3-objects.s3.localhost.localstack.cloud:4566', 'amz-sdk-invocation-id': 'ebe65d10-9ef6-1702-7568-d964262f3905', 'amz-sdk-request': 'attempt=1; max=4', 'Authorization': 'AWS4-HMAC-SHA256 Credential=arbitrary-value-is-required-for-localstack/20231005/ap-south-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;host;x-amz-content-sha256;x-amz-date;x-amz-te, Signature=bbb772c4b64742d6a3c7f2129128cf56191507fecfb15da8ea0f9aac27125dd5', 'User-Agent': 'aws-sdk-java/2.20.159 Linux/6.4.6-76060406-generic OpenJDK_64-Bit_Server_VM/17.0.7+7-LTS Java/17.0.7 scala/2.12.10 kotlin/1.9.10-release-459 vendor/Amazon.com_Inc. io/async http/NettyNio cfg/retry-mode/legacy', 'x-amz-content-sha256': 'UNSIGNED-PAYLOAD', 'X-Amz-Date': '20231005T221850Z', 'x-amz-te': 'append-md5', 'x-localstack-tgt-api': 's3'}); Response(StreamingBody(unknown), headers={'Content-Type': 'application/octet-stream', 'accept-ranges': 'bytes', 'Last-Modified': 'Thu, 05 Oct 2023 22:18:50 GMT', 'ETag': '"fbaff7b2f613227d604f3e89f517d379"', 'x-amz-version-id': '55460f66-2508-4ad5-9404-08a6b97b5e34', 'x-amz-tagging-count': '3', 'x-amz-request-id': '0a2a9574-467d-40f9-9cd4-d203c3eceb2c', 'x-amz-id-2': 's9lzHYrFp76ZVxRcpX9+5cjAnEH2ROuNkd2BHfIa6UkFVdtjf5mKR3/eTPFvsiP/XV/VLi31234=', 'Connection': 'close'})
```

If that's the right part, I see three separate headers. Two `Content-Length` (one of which is 0) and `ContentLength`.

**Comment by dominic-jones** [comment]
> I can see in the diff between 2.1 and 2.2 that the URL to access LocalStack has changed too, and I think with 2.2 and 2.3 if you access LocalStack S3 with http://localhost:4566 or http://s3.localhost.localstack.cloud:4566 and accessing the bucket in "path style" (bucket part as the path and not the host), you shouldn't encounter the issue with ContentLength missing. This is indeed only a workaround, and we will work to fix this issue. Thanks!

Okay I clearly checked this too late. I thought you were recommending NOT to use path style access, whereas you said in fact to try it.

If I use `.forcePathStyle(true)` (or `.serviceConfiguration { it.pathStyleAccessEnabled(true) }`) on my S3Client, then yes it works and I get a content-length returned correctly.

**Comment by dominic-jones** [comment]
My apologies. Having rerun the tests in more detail today, this appears to be successful with 2.3.0 after all. This bug perhaps only affects users of 2.2.0.

**Comment by bentsku** [comment]
> Naturally not a long term solution as this is a deprecated approach.

Yes, that's true, I'm not seeing AWS moving away soon, but they might surprise us! I'm thinking about changing our approach to the virtual hosted requests, which should fix these kind of issues. 

I'm also sorry for the misunderstanding, hope it didn't make you lose too much time. 

Well, I'm glad your issue is fixed! We might have had a regression for a while that would have been fixed with the multiple PRs like the one I've linked above. 

Don't hesitate to re-open if you encounter any issue again, or another issue for something different. Thanks again for the report!
